CMakeLists.txt

```diff
@@ -60,49 +67,71 @@ else()
+#     ERROR_VARIABLE  CLANG_RESOURCE_DIR_ERROR
+#     OUTPUT_STRIP_TRAILING_WHITESPACE)
+# 
+# if(CLANG_RESOURCE_DIR_RESULT)
+#     message(FATAL_ERROR "Could not locate the resource dir for clang we are going to link against")
+#     message(FATAL_ERROR ${CLANG_RESOURCE_DIR_ERROR})
+# endif()
 
 message(STATUS "Using the following resource dir for clang: ${CLANG_RESOURCE_DIR}")
 
 file(GLOB EMITTER_FILES CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/emitters/*.cpp)
 file(GLOB MATCHER_FILES CONFIGURE_DEPENDS ${PROJECT_SOURCE_DIR}/matchers/*.cpp)
 file(GLOB SRC_FILES CONFIGURE_DEPENDS ${PROJECT_SOURCE_DIR}/sources/*.cpp)
-
 file(GLOB YAML_CPP_SRC_FILES CONFIGURE_DEPENDS ${PROJECT_SOURCE_DIR}/submodules/yaml-cpp/src/*.cpp)
 
 #TODO: use target_source
-add_executable(hyde  ${EMITTER_FILES} ${MATCHER_FILES} ${SRC_FILES} ${YAML_CPP_SRC_FILES})
+add_executable(hyde
+    ${EMITTER_FILES}
+    ${MATCHER_FILES}
+    ${SRC_FILES}
+    ${YAML_CPP_SRC_FILES}
+)
+
+target_compile_definitions(hyde
+    PRIVATE
+        CLANG_RESOURCE_DIR="${CLANG_RESOURCE_DIR}"
+        LLVM_INCLUDE_DIR="${llvm_INCLUDE_DIR}"
+)
+
+target_include_directories(hyde
+    PUBLIC
+        ${Boost_INCLUDE_DIRS}
+        ${CLANG_INCLUDE_DIRS}
+        ${CMAKE_CURRENT_SOURCE_DIR}
+        ${CMAKE_CURRENT_SOURCE_DIR}/include
+        ${LLVM_INCLUDE_DIRS}
+        ${PROJECT_SOURCE_DIR}/submodules/yaml-cpp/include/
+        ${PROJECT_SOURCE_DIR}/submodules/json/include/
+)
+
+target_compile_options(hyde
+    PUBLIC
+        -Wall
+        -Wno-comment
+        -Werror
+        -Wno-range-loop-analysis
+        -DHYDE_FORCE_BOOST_FILESYSTEM=1
+)
+
 if (NOT LLVM_ENABLE_RTTI)
-    target_compile_options (hyde PRIVATE -fno-rtti)
+    target_compile_options(hyde PRIVATE -fno-rtti)
 endif()
 
-target_compile_definitions(hyde PRIVATE CLANG_RESOURCE_DIR="${CLANG_RESOURCE_DIR}" )
-
-target_compile_definitions(hyde PRIVATE LLVM_INCLUDE_DIR="${LLVM_INCLUDE_DIR}" )
-
-target_include_directories(hyde PUBLIC ${Boost_INCLUDE_DIRS})
-target_include_directories(hyde PUBLIC ${CLANG_INCLUDE_DIRS})
-target_include_directories(hyde PUBLIC ${CMAKE_CURRENT_SOURCE_DIR})
-target_include_directories(hyde PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
-target_include_directories(hyde PUBLIC ${LLVM_INCLUDE_DIRS})
-target_include_directories(hyde PUBLIC ${PROJECT_SOURCE_DIR}/submodules/yaml-cpp/include/)
-target_include_directories(hyde PUBLIC ${PROJECT_SOURCE_DIR}/submodules/json/include/)
-
-target_compile_options(hyde PUBLIC -Wall -Wno-comment -Werror -Wno-range-loop-analysis -DHYDE_FORCE_BOOST_FILESYSTEM=1)
-
 target_link_libraries(hyde
-                      ${Boost_FILESYSTEM_LIBRARY}
-                      ${Boost_SYSTEM_LIBRARY}
-                      clangASTMatchers
-                      clangBasic
-                      clangTooling
-                      )
+    ${Boost_FILESYSTEM_LIBRARY}
+    ${Boost_SYSTEM_LIBRARY}
+    clang
+    clangASTMatchers
+    clangBasic
+    clangTooling
```

Comment:
Looking at the include directives I think there are a few more direct dependencies that should be listed here: `clangLex`, `clangFrontend` etc.

Reply:
Is it worth adding them even if hyde doesn't need them?

Reply:
This is the list of include that I saw in the `hyde` source files:
```
#include "clang/AST/...
#include "clang/ASTMatchers/...
#include "clang/Frontend/...
#include "clang/Lex/...
#include "clang/Tooling/...
```
so I assumed it depends on all these components.

It is possible that 1/ the dependency is header-only and doesn't require linking or 2/ the dependency is pulled by another dependency already listed, but relying on 1/ is brittle because it can easily break if some definition is moved to a .cpp at some point and 2/ is brittle because it can break if you remove some unrelated dependency or if the dependency is updated to no longer depends on the one that you were pulling implicitly. This is basically a variation of `include-what-you-use`: `target-link-what-you-use` ^^

Reply:
Fair enough, thanks for the explanation. I'll add them.